src/core/tsi/ssl_transport_security.cc

```diff
@@ -490,21 +490,34 @@ static tsi_result peer_from_x509(X509* cert, int include_certificate_type,
   return result;
 }
 
-/* Logs the SSL error stack. */
-static void log_ssl_error_stack(void) {
-  unsigned long err;
-  while ((err = ERR_get_error()) != 0) {
-    char details[256];
-    ERR_error_string_n(static_cast<uint32_t>(err), details, sizeof(details));
-    gpr_log(GPR_ERROR, "%s", details);
+/* Clears all SSL errors from the stack, and returns them as a string. */
+static std::string ssl_error_stack_to_string(void) {
+  BIO* bio = BIO_new(BIO_s_mem());
+  ERR_print_errors(bio);
+
+  const uint8_t* buf;
+  size_t buf_len;
+  std::string result;
+  if (!BIO_mem_contents(bio, &buf, &buf_len) || buf_len == 0) {
+    result = "error could not be retrieved from the SSL stack";
+  } else {
+    result = std::string(reinterpret_cast<const char*>(buf), buf_len);
   }
+  BIO_free(bio);
+  return result;
 }
 
 /* Performs an SSL_read and handle errors. */
 static tsi_result do_ssl_read(SSL* ssl, unsigned char* unprotected_bytes,
                               size_t* unprotected_bytes_size) {
-  int read_from_ssl;
   GPR_ASSERT(*unprotected_bytes_size <= INT_MAX);
+  // Gracefully handle the case where some other part of the application left
+  // an error on the stack.
+  if (ERR_peek_error() != 0) {
+    gpr_log(GPR_ERROR, "do_ssl_read invoked with a dirty SSL error stack: %s",
+            ssl_error_stack_to_string().c_str());
```

Comment:
So adding a warning here only makes sense if gRPC believes it is a bug to for other code to handle an OpenSSL error without consuming the error queue. Otherwise you'll end up causing a lot of noise.

It is true that leaving the error queue dirty will result in some weirdness down the line. At the same time, I think this is a lot of fuss. It's also a little odd to have `ssl_error_stack_to_string` return a string with newlines. I suggest either:

Option 1: say that leaving things in the error queue isn't a bug worth warning on and just replace all these `if (ERR_peek_error() ...) { ... }` blocks with a single `ERR_clear_error()` call.

Option 2: say we want to warn about leaving things in the error queue. In that case, let's just call the log function twice like this:

```
static void ensure_empty_error_queue(const char* function) {
  if (ERR_peek_error() != 0) {
    gpr_log(GPR_ERROR, "%s invoked with a dirty OpenSSL error queue.", function);
    log_ssl_error_queue();
  }
}
```

Reason for the helper function is...

Reply:
I think you've convinced me that I'm being too fussy and (1.) is correct.  Handling the other cases as well.